File: python/src/threads/loop/gates.py
"""Hook points around a turn: before_input, before_model, after_model,
on_stop and stop_failure. Each gate's decisions are recorded, in the same batch as what they
lead to, before the work they gate. Whether a gate already ran is read from the log, so a
resumed run never asks a hook twice (replay reads recorded decisions)."""

from collections.abc import Mapping, Sequence
from typing import Final, Literal

from pydantic import JsonValue, TypeAdapter, ValidationError

from threads.hooks.runner import (
    INPUT,
    MODEL,
    NOTHING,
    RESPONSE,
    STOP,
    TEXTS,
    Ran,
    decision_draft,
    injected,
)
from threads.hooks.types import HookName, Source, wire_name
from threads.log import (
    Event,
    HookDecisionEvent,
    InjectedEvent,
    ModelRequestEvent,
    UserInputEvent,
)
from threads.loop.drafts import call_draft, draft
from threads.loop.history import last_response, response_calls, turn_events
from threads.loop.results import As, result_draft
from threads.loop.runtime import FAILED_CODES, Failed, Halt, Runtime, lost
from threads.result import Err
from threads.store import Draft

AGAIN: Final = "again"
type Gated = Halt | Literal["again"] | None
"""None: the gate passed; AGAIN: it appended, so the loop decides again; a Halt stops."""

MAX_RETRIES: Final = 2
"""after_model retry rounds per turn before a retry is a deny."""
MAX_STOP_CONTINUES: Final = 3
"""on_stop continuations per turn before stop_hook_limit."""
_STOPPED: Final = "on_stop"
_FIELDS: Final = TypeAdapter[Mapping[str, object]](Mapping[str, object])
_TEXTS: Final = TypeAdapter[list[str]](list[str])


def decided(events: Sequence[Event], hook: HookName, key: str, value: str) -> bool:
    """A decision of `hook` names this call, request or input."""
    return any(
        isinstance(e, HookDecisionEvent)
        and e.data.hook == wire_name(hook)
        and getattr(e.data, key) == value
        for e in events
    )


def _fields(ran: Ran[object]) -> Mapping[str, object]:
    try:
        return _FIELDS.validate_python(ran.value)
    except ValidationError:
        return {}


def verdict(ran: Ran[object], fail: str = "deny") -> str:
    """The decision a hook's answer stands for; a failure stands for `fail`."""
    decision = _fields(ran).get("decision")
    return decision if ran.failure is None and isinstance(decision, str) else fail


def said(ran: Ran[object], field: str) -> str | None:
    value = _fields(ran).get(field)
    return value if isinstance(value, str) else None


def texts(ran: Ran[object], field: str = "injections") -> list[str]:
    """Injections on an allow or proceed, or a hook's plain list of texts."""
    fields = _fields(ran)
    raw = fields.get(field, []) if fields else ran.value
    try:
        return _TEXTS.validate_python(raw)
    except ValidationError:
        return []


async def append(rt: Runtime, drafts: Sequence[Draft]) -> Gated:
    done = await rt.append(*drafts)
    return lost(done.error) if isinstance(done, Err) else AGAIN


async def before_request(rt: Runtime) -> Gated:
    """before_input for every input of the turn not yet decided, then before_model once per
    attempt."""
    turn = turn_events(rt.events)
    if rt.hooks.has("before_input"):
        for event in turn:
            if isinstance(event, UserInputEvent) and not decided(
                turn, "before_input", "input_event_id", event.event_id
            ):
                return await _before_input(rt, event)
    if rt.hooks.has("before_model") and not _model_gated(turn):
        return await _before_model(rt)
    return None


async def _before_input(rt: Runtime, event: UserInputEvent) -> Gated:
    """A deny keeps the input for audit, renders it in no request, and ends the turn
    input_denied."""
    ran = await rt.hooks.run("before_input", INPUT, event.data)
    denied = any(verdict(r) != "allow" for r in ran)
    ids = {"input_event_id": event.event_id}
    drafts = [decision_draft("before_input", r, verdict(r), said(r, "reason"), **ids) for r in ran]
    if denied:
        drafts.append(draft("turn_completed", {"reason": "input_denied"}))
    else:
        drafts += [d for r in ran for d in injected(r.extension, texts(r))]
    return await append(rt, drafts)


def _model_gated(turn: Sequence[Event]) -> bool:
    """before_model already decided this attempt: after the turn's latest turn request."""
    since = 0
    for index, event in enumerate(turn):
        if isinstance(event, ModelRequestEvent) and event.data.purpose != "compaction":
            since = index + 1
    return any(
        isinstance(e, HookDecisionEvent) and e.data.hook == "before_model" for e in turn[since:]
    )


async def _before_model(rt: Runtime) -> Gated:
    """A deny means the attempt is never sent; the turn ends error."""
    ran = await rt.hooks.run("before_model", MODEL, rt.writer.state())
    drafts = [decision_draft("before_model", r, verdict(r), said(r, "reason")) for r in ran]
    if any(verdict(r) != "proceed" for r in ran):
        drafts.append(draft("turn_completed", {"reason": "error"}))
    else:
        drafts += [d for r in ran for d in injected(r.extension, texts(r))]
    return await append(rt, drafts)


async def after_model(rt: Runtime) -> Gated:
    """Gates the response's undispatched calls and the release of its output.
    deny closes them denied and ends the turn with the output withheld; guide closes them and
    re-asks with a trusted instruction. ponytail: retry is treated as guide(reason): a fresh
    attempt that hides the recorded response needs a render rule the spec doesn't have yet."""
    turn = turn_events(rt.events)
    response = last_response(turn)
    if not rt.hooks.has("after_model") or response is None:
        return None
    request = response.data.request_event_id
    if decided(turn, "after_model", "request_event_id", request):
        return None
    ran = await rt.hooks.run("after_model", RESPONSE, rt.writer.state(), response.data)
    ids = {"request_event_id": request}
    drafts = [
        decision_draft("after_model", r, verdict(r), said(r, "reason") or said(r, "text"), **ids)
        for r in ran
    ]
    verdicts = {verdict(r) for r in ran}
    if verdicts == {"proceed"}:
        return await append(rt, drafts)
    # Retries are capped per turn, counted from the log so recovery can't reset them.
    denied = "deny" in verdicts or ("retry" in verdicts and _retries(turn) >= MAX_RETRIES)
    why = "denied by an output guardrail" if denied else "withheld: the response was guided"
    how = As("denied", True, "host")
    # An older writer recorded the calls with the response: they are pending already.
    for call_id in rt.fold.pending:
        drafts.append(await result_draft(rt, call_id, why, how))
    for use, call in response_calls(rt.events):
        if call is None:
            drafts += [call_draft(request, use), await result_draft(rt, use.call_id, why, how)]
    if denied:
        drafts.append(draft("turn_completed", {"reason": "error"}))
    else:
        guide = "\n".join(said(r, "text") or said(r, "reason") or "" for r in ran)
        drafts.append(_instruction("after_model", guide))
    return await append(rt, drafts)


def _retries(turn: Sequence[Event]) -> int:
    """after_model rounds that asked for a retry: one per gated response."""
    return len(
        {
            e.data.request_event_id
            for e in turn
            if isinstance(e, HookDecisionEvent)
            and e.data.hook == "after_model"
            and e.data.decision == "retry"
        }
    )


def _instruction(origin: str, text: str) -> Draft:
    data: dict[str, JsonValue] = {
        "source": "hook",
        "trust": "trusted_instruction",
        "origin": {"id": origin},
        "text": text,
    }
    return draft("injected", data)


async def end_turn(rt: Runtime) -> Halt | None:
    """The turn would end: on_stop may continue it, at most MAX_STOP_CONTINUES times; a failed
    on_stop stops."""
    if not rt.hooks.has("on_stop"):
        return await _done(rt, "end_turn")
    ran = await rt.hooks.run("on_stop", STOP, rt.writer.state())
    drafts = [decision_draft("on_stop", r, verdict(r, "stop"), said(r, "reason")) for r in ran]
    reasons = [said(r, "reason") or "" for r in ran if verdict(r, "stop") == "continue"]
    if not reasons:
        drafts.append(draft("turn_completed", {"reason": "end_turn"}))
    elif _continues(rt.events) >= MAX_STOP_CONTINUES:
        drafts.append(draft("turn_completed", {"reason": "stop_hook_limit"}))
    else:
        drafts.append(_instruction(_STOPPED, "\n".join(reasons)))
    done = await rt.append(*drafts)
    return lost(done.error) if isinstance(done, Err) else None


def _continues(events: Sequence[Event]) -> int:
    """Rounds of on_stop that continued the turn: one instruction each."""
    return sum(
        1
        for e in turn_events(events)
        if isinstance(e, InjectedEvent) and e.data.source == "hook" and e.data.origin.id == _STOPPED
    )


async def _done(rt: Runtime, reason: str) -> Halt | None:
    done = await rt.append(draft("turn_completed", {"reason": reason}))
    return lost(done.error) if isinstance(done, Err) else None


async def failed(rt: Runtime, reason: str) -> Halt | None:
    """stop_failure observers when a turn ends in a failure; on_stop doesn't run then."""
    code = FAILED_CODES.get(reason)
    return None if code is None else await observe(rt, "on_stop_failure", code)


async def observe(rt: Runtime, hook: HookName, *args: object, **ids: str) -> Halt | None:
    """An observe hook: it can't change execution, and only its failure is recorded."""
    if not rt.hooks.has(hook):
        return None
    ran = await rt.hooks.run(hook, NOTHING, *args)
    drafts = [decision_draft(hook, r, "failed", **ids) for r in ran if r.failure is not None]
    if not drafts:
        return None
    done = await rt.append(*drafts)
    return lost(done.error) if isinstance(done, Err) else None


async def session_start(rt: Runtime, source: Source) -> Halt | None:
    """Context hook: its injections go in before the input; a failure denies the session, so
    the run takes no input."""
    if not rt.hooks.has("session_start"):
        return None
    ran = await rt.hooks.run("session_start", TEXTS, source)
    drafts = [decision_draft("session_start", r, "proceed") for r in ran]
    failures = [f"{r.extension}: {r.failure}" for r in ran if r.failure is not None]
    if not failures:
        drafts += [d for r in ran for d in injected(r.extension, texts(r))]
    done = await rt.append(*drafts)
    if isinstance(done, Err):
        return lost(done.error)
    return (
        Failed("input_denied", f"session_start failed: {'; '.join(failures)}") if failures else None
    )

File: typescript/packages/core/src/loop/gates.ts
import { assertNever } from "../assert-never";
import type { EventOf } from "../fold/state";
import type { Outcome } from "../hooks/invoke";
import type { HookReturns } from "../hooks/types";
import type { KnownEvent } from "../log";
import type { EventDraft } from "../store";
import { draft } from "./drafts";
import { context, decision, defining, injection, recorded, run } from "./hooks";
import type { Session } from "./session";
import { stepEvents, turnEvents } from "./turn";
import type { Halt } from "./types";

// The gates in front of a turn request: the input guardrail, the
// tool-output guardrail, after_tool_batch and before_model. Each decision is recorded with
// what it causes in one append, before the request that it gates, so a recovered run reads the
// decision and never runs the hook again.

/** The gate ended the turn; the loop moves on to the next step. */
export type Gated = Halt | "ended" | undefined;

const encoder = new TextEncoder();

/** before_input on the turn's input, before the turn's first request. */
export async function inputGate(s: Session): Promise<Gated> {
  const turn = turnEvents(s.events, s.fold);
  const input = turn[0];
  if (input?.type !== "user_input") return undefined;
  if (turn.some((e) => e.type === "model_request")) return undefined;
  const key = { input_event_id: input.event_id };
  for (const ext of defining(s, "before_input")) {
    if (recorded(turn, ext.name, "before_input", key) !== undefined) continue;
    const out = await run(ext, "before_input", [input.data]);
    const refused = refusal(out);
    if (refused !== undefined) {
      const { decided, reason } = refused;
      return (
        (await s.append(
          decision(ext.name, "before_input", decided, key, reason),
          draft.turnCompleted("input_denied"),
        )) ?? "ended"
      );
    }
    if (out.kind === "failed" || out.value.decision === "deny") continue;
    const stopped = await s.append(
      decision(ext.name, "before_input", "allow", key),
      ...(out.value.injections ?? []).map((t) => injection(ext.name, t)),
    );
    if (stopped !== undefined) return stopped;
  }
  return undefined;
}

/**
 * before_tool_result on each executed result of this step. redact and deny change only what
 * later requests render (context_edited{guardrail}); the raw result stays in the log.
 */
export async function resultsGate(s: Session): Promise<Halt | undefined> {
  const step = stepEvents(s.events, s.fold);
  const results = step.filter(
    (e): e is EventOf<"tool_result"> =>
      e.type === "tool_result" && e.data.origin === "executed",
  );
  for (const result of results) {
    const key = { call_id: result.data.call_id };
    const call = s.events.findLast(
      (e) => e.type === "tool_call" && e.data.call_id === key.call_id,
    );
    if (call?.type !== "tool_call") continue;
    for (const ext of defining(s, "before_tool_result")) {
      if (recorded(step, ext.name, "before_tool_result", key) !== undefined)
        continue;
      const out = await run(
        ext,
        "before_tool_result",
        [call.data, result.data],
        key.call_id,
      );
      const stopped = await s.append(...guarded(ext.name, result, out));
      if (stopped !== undefined) return stopped;
    }
  }
  return undefined;
}

type ResultOutcome =
  | { readonly kind: "ok"; readonly value: HookReturns["before_tool_result"] }
  | { readonly kind: "failed"; readonly reason: string };

function guarded(
  ext: string,
  result: EventOf<"tool_result">,
  out: ResultOutcome,
): readonly EventDraft[] {
  const key = { call_id: result.data.call_id };
  const clear = (decided: "deny" | "failed", reason: string) => [
    decision(ext, "before_tool_result", decided, key, reason),
    draft.contextEdited({
      reason: "guardrail",
      edits: [{ call_id: key.call_id, action: "clear" }],
    }),
  ];
  if (out.kind === "failed") return clear("failed", out.reason);
  const value = out.value;
  switch (value.decision) {
    case "proceed":
      return [decision(ext, "before_tool_result", "proceed", key)];
    case "deny":
      return clear("deny", value.reason);
    case "redact": {
      const part = textPart(result);
      if (part === undefined || !value.spans.every((x) => inside(part.text, x)))
        return clear("failed", "redaction spans outside the result's text");
      return [
        decision(ext, "before_tool_result", "redact", key),
        draft.contextEdited({
          reason: "guardrail",
          edits: [
            {
              call_id: key.call_id,
              action: "redact",
              part: part.index,
              spans: [...value.spans],
            },
          ],
        }),
      ];
    }
    default:
      return assertNever(value);
  }
}

/** The result's first text part: the preview when it has no content (render rule). */
function textPart(
  result: EventOf<"tool_result">,
): { readonly index: number; readonly text: string } | undefined {
  const { content, preview } = result.data;
  if (content === undefined)
    return preview === undefined ? undefined : { index: 0, text: preview };
  const index = content.findIndex((p) => p.type === "text");
  const part = content[index];
  return part?.type === "text" ? { index, text: part.text } : undefined;
}

/** A span inside the text's UTF-8 bytes, on character boundaries (semantic rule 19). */
function inside(text: string, span: { start: number; end: number }): boolean {
  const bytes = encoder.encode(text);
  const boundary = (at: number): boolean =>
    at === bytes.length || ((bytes[at] ?? 0) & 0xc0) !== 0x80;
  return (
    span.start >= 0 &&
    span.start < span.end &&
    span.end <= bytes.length &&
    boundary(span.start) &&
    boundary(span.end)
  );
}

/** after_tool_batch once every result of the last response is recorded; a failure ends the turn. */
export async function batchGate(s: Session): Promise<Gated> {
  const step = stepEvents(s.events, s.fold);
  if (!step.some((e) => e.type === "tool_result")) return undefined;
  // Keyed by the response whose calls this batch answered, so the decision names its subject and
  // a re-run of the step reads its own record back (parity with Python's after_batch).
  const response = turnEvents(s.events, s.fold).findLast(
    (e) => e.type === "model_response",
  );
  if (response?.type !== "model_response") return undefined;
  const key = { request_event_id: response.data.request_event_id };
  const got = await context(s, "after_tool_batch", [s.state()], step, key);
  if (got === "failed")
    return (await s.append(draft.turnCompleted("error"))) ?? "ended";
  return got;
}

/** before_model per attempt: proceed with its injections, or the attempt isn't sent. */
export async function modelGate(s: Session): Promise<Gated> {
  const window = sinceLastRequest(stepEvents(s.events, s.fold));
  for (const ext of defining(s, "before_model")) {
    if (recorded(window, ext.name, "before_model") !== undefined) continue;
    const out = await run(ext, "before_model", [s.state()]);
    const refused = refusal(out);
    if (refused !== undefined) {
      const { decided, reason } = refused;
      return (
        (await s.append(
          decision(ext.name, "before_model", decided, {}, reason),
          draft.turnCompleted("error"),
        )) ?? "ended"
      );
    }
    if (out.kind === "failed" || out.value.decision === "deny") continue;
    const stopped = await s.append(
      decision(ext.name, "before_model", "proceed"),
      ...(out.value.injections ?? []).map((t) => injection(ext.name, t)),
    );
    if (stopped !== undefined) return stopped;
  }
  return undefined;
}

/** A gate's refusal: the hook failed (it denies) or denied. */
function refusal(
  out: Outcome<"before_input" | "before_model">,
):
  | { readonly decided: "failed" | "deny"; readonly reason: string }
  | undefined {
  if (out.kind === "failed") return { decided: "failed", reason: out.reason };
  return out.value.decision === "deny"
    ? { decided: "deny", reason: out.value.reason }
    : undefined;
}

function sinceLastRequest(step: readonly KnownEvent[]): readonly KnownEvent[] {
  return step.slice(step.findLastIndex((e) => e.type === "model_request") + 1);
}
